lib/littlefs: Update LittleFS to v2.8.0.

DvdGiessen opened the pull request:
This updates LittleFS from 2.5.0 to 2.8.0. It brings a few small performance improvements, and most notably an improvement that fixes "best case fixable corruption, worst case significantly reduced data retention" when power failure occurs during a write.

For a detailed list of changes, see [the LittleFS releases page](https://github.com/littlefs-project/littlefs/releases).

One thing to note is that this version also bumps the on-disk version, meaning that a filesystem written by this version can no longer be read by an older MicroPython / LittleFS version.

I've been running on the latest LittleFS version on my boards for a while without any issues.

DvdGiessen committed:
lib/littlefs: Update LittleFS to v2.8.1.

Signed-off-by: Daniël van de Giessen <daniel@dvdgiessen.nl>

dpgeorge commented:
Thanks for doing this, it's definitely a good step forward.

> One thing to note is that this version also bumps the on-disk version, meaning that a filesystem written by this version can no longer be read by an older MicroPython / LittleFS version.

But, can this new version read older on-disk filesystems?

The increase in code size is quite a bit.  Do you have any idea why, is it simply due to lots of small improvements, each which costs a little bit of code size?

The CI is failing, because of some variable shadowing errors.  Would be good to report that upstream to littlefs, and in the meantime add special CFLAGS for those files which have the issue.

DvdGiessen commented:
> But, can this new version read older on-disk filesystems?

Yes, it can read older on-disk versions, I've verified this by flashing an older version of the filesystem on an ESP32-S3 and mounting it with this new version without issues.

Upon the first write to the filesystem it will also update the on-disk version field, so filesystems only become unusable by older versions after a newer version has written to it.

> The increase in code size is quite a bit. Do you have any idea why, is it simply due to lots of small improvements, each which costs a little bit of code size?

 I've glanced over the map diff, think it's a number of small changes across the code and a few new functions:

```diff
+lfs2_fs_gc                                        esp-idf/main_esp32s3/libmain_esp32s3.a(lfs2.c.obj)
+lfs2_fs_grow                                      esp-idf/main_esp32s3/libmain_esp32s3.a(lfs2.c.obj)
+lfs2_fs_mkconsistent                              esp-idf/main_esp32s3/libmain_esp32s3.a(lfs2.c.obj)
+lfs2_fs_rawgrow                                   esp-idf/main_esp32s3/libmain_esp32s3.a(lfs2.c.obj)
+lfs2_fs_rawmkconsistent                           esp-idf/main_esp32s3/libmain_esp32s3.a(lfs2.c.obj)
 lfs2_fs_size                                      esp-idf/main_esp32s3/libmain_esp32s3.a(lfs2.c.obj)
+lfs2_fs_stat                                      esp-idf/main_esp32s3/libmain_esp32s3.a(lfs2.c.obj)
```

The increase in size is indeed quite a bit. We could try to move additional functionality behind a compile flag and try to upstream those changes?

> The CI is failing, because of some variable shadowing errors. Would be good to report that upstream to littlefs, and in the meantime add special CFLAGS for those files which have the issue.

I've added the special CFLAGS for the affected file. Note I've also removed `-Wno-missing-field-initializers`, which appears to be solved upstream in https://github.com/littlefs-project/littlefs/pull/405.

There's already an open issue upstream for the shadowing (https://github.com/littlefs-project/littlefs/pull/873); I'll open an issue or PR for the other warning.

dpgeorge commented:
It looks like 2.8.1 was just released, so probably worth updating this PR to that version, will hopefully fix at least one of the above warnings.

dpgeorge commented:
Regarding size: it looks like most of the increase comes from v2.6, which bumped the on-disk version, and had quite a few improvements and bug fixes.  The backwards-compatibility support for the previous on-disk version probably adds a bit of code size as well.

And then subsequent releases v2.7 and v2.8 added about 1.7% to the code size, which is maybe 250 bytes more.

For the additional functions that we don't use, we are not paying for them due to the linker GC.  So the size increase is mostly from performance and bug fixes.

I checked mboot which builds lfs2 in read-only mode, and it increases by +140 bytes (stm32).  That's acceptable.

One thing to note with the bump to the on-disk version: if your board has mboot with lfs2 support then it will need updating if the main application is updated.  Because the main application will write to the filesystem and that will change the version, and then the old mboot won't be able to read the filesystem anymore.  We don't have any boards that build mboot with lfs2 by default, so that's only an issue for those who define custom boards with lfs2 enabled.

Summary: I think this is a good update and we should do it, and we can't do much about the code size increase.
